File: src/demo2_apriltag_docking/demo2_apriltag_docking/tag_policy.py
from dataclasses import dataclass
import math


@dataclass(frozen=True)
class DockSpec:
    tag_id: int
    dock_id: str
    dock_type: str
    tag_frame: str


@dataclass(frozen=True)
class Detection:
    tag_id: int
    hamming: int
    decision_margin: float
    stamp: float
    x: float
    y: float
    yaw: float


@dataclass(frozen=True)
class GateResult:
    accepted: bool
    reason: str
    detection: Detection | None = None
    dock: DockSpec | None = None


def load_dock_specs(raw):
    docks = raw.get('docks') if isinstance(raw, dict) else None
    if not isinstance(docks, dict) or not docks:
        raise ValueError('docks must be a non-empty mapping')

    specs = {}
    dock_ids = set()
    for raw_tag_id, value in docks.items():
        try:
            tag_id = int(raw_tag_id)
        except (TypeError, ValueError) as exc:
            raise ValueError(f'invalid tag id: {raw_tag_id!r}') from exc

        if tag_id < 0 or tag_id in specs or not isinstance(value, dict):
            raise ValueError(f'invalid or duplicate tag id: {raw_tag_id!r}')

        fields = [value.get(name) for name in ('dock_id', 'dock_type', 'tag_frame')]
        if any(not isinstance(field, str) or not field.strip() for field in fields):
            raise ValueError(f'tag {tag_id} has missing or empty fields')

        dock_id, dock_type, tag_frame = fields
        if dock_id in dock_ids:
            raise ValueError(f'duplicate dock id: {dock_id}')

        specs[tag_id] = DockSpec(tag_id, dock_id, dock_type, tag_frame)
        dock_ids.add(dock_id)

    return specs


class TagGate:
    def __init__(
        self,
        *,
        specs,
        min_margin,
        max_hamming,
        confirmations,
        confirmation_window,
        publish_period,
        loss_timeout,
        max_translation_jump,
        max_yaw_jump,
    ):
        self.specs = specs
        self.min_margin = min_margin
        self.max_hamming = max_hamming
        self.confirmations = confirmations
        self.confirmation_window = confirmation_window
        self.publish_period = publish_period
        self.loss_timeout = loss_timeout
        self.max_translation_jump = max_translation_jump
        self.max_yaw_jump = max_yaw_jump

        self._active_tag_id = None
        self._confirmation_count = 0
        self._confirmation_started = None
        self._confirmed = False
        self._last_accepted = None
        self._last_publication = None
        self._last_seen = None

    def evaluate(self, detections, now):
        if not detections:
            return self._reject('NO_TAG')
        if len(detections) != 1:
            return self._reject('MULTI_TAG')

        detection = detections[0]
        dock = self.specs.get(detection.tag_id)
        if dock is None:
            return self._reject('UNKNOWN_TAG')
        if detection.hamming > self.max_hamming:
            return self._reject('HAMMING')
        if detection.decision_margin < self.min_margin:
            return self._reject('LOW_MARGIN')

        if (
            self._last_seen is not None
            and now - self._last_seen > self.loss_timeout
        ):
            self._last_accepted = None
            self._reset_confirmation()
        self._last_seen = now
        if self._is_pose_jump(detection):
            return self._reject('POSE_JUMP')

        if not self._confirmed:
            expired = (
                self._confirmation_started is None
                or now - self._confirmation_started > self.confirmation_window
            )
            if self._active_tag_id != detection.tag_id or expired:
                self._active_tag_id = detection.tag_id
                self._confirmation_count = 1
                self._confirmation_started = now
            else:
                self._confirmation_count += 1

            if self._confirmation_count < self.confirmations:
                return GateResult(False, 'CONFIRMING', detection, dock)
            self._confirmed = True

        if (
            self._last_publication is not None
            and now - self._last_publication < self.publish_period
        ):
            return GateResult(False, 'RATE_LIMITED', detection, dock)

        self._last_accepted = detection
        self._last_publication = now
        return GateResult(True, 'ACCEPTED', detection, dock)

    def loss_reason(self, now):
        if self._last_seen is None or now - self._last_seen <= self.loss_timeout:
            return None
        self._last_seen = None
        self._last_accepted = None
        self._reset_confirmation()
        return 'TAG_LOST'

    def _reject(self, reason):
        self._reset_confirmation()
        return GateResult(False, reason)

    def _reset_confirmation(self):
        self._active_tag_id = None
        self._confirmation_count = 0
        self._confirmation_started = None
        self._confirmed = False

    def _is_pose_jump(self, detection):
        previous = self._last_accepted
        if previous is None or previous.tag_id != detection.tag_id:
            return False

        translation = math.hypot(detection.x - previous.x, detection.y - previous.y)
        yaw_delta = (detection.yaw - previous.yaw + math.pi) % (2.0 * math.pi) - math.pi
        return (
            translation > self.max_translation_jump
            or abs(yaw_delta) > self.max_yaw_jump
        )

File: src/demo2_apriltag_docking_cpp/src/core/tag_policy.cpp
#include "demo2_apriltag_docking_cpp/core/tag_policy.hpp"

#include <cmath>
#include <stdexcept>
#include <utility>

namespace demo2_apriltag_docking_cpp::core {
namespace {

constexpr double kPi = 3.141592653589793238462643383279502884;
constexpr double kTwoPi = 2.0 * kPi;

bool finite(double value)
{
  return std::isfinite(value);
}

double wrapped_yaw_delta(double value)
{
  double wrapped = std::fmod(value + kPi, kTwoPi);
  if (wrapped < 0.0) {
    wrapped += kTwoPi;
  }
  return wrapped - kPi;
}

}  // namespace

TagGate::TagGate(
  std::unordered_map<int, DockSpec> specs,
  double min_margin,
  int max_hamming,
  int confirmations,
  double confirmation_window,
  double publish_period,
  double loss_timeout,
  double max_translation_jump,
  double max_yaw_jump)
: specs_(std::move(specs)),
  min_margin_(min_margin),
  max_hamming_(max_hamming),
  confirmations_(confirmations),
  confirmation_window_(confirmation_window),
  publish_period_(publish_period),
  loss_timeout_(loss_timeout),
  max_translation_jump_(max_translation_jump),
  max_yaw_jump_(max_yaw_jump)
{
  if (specs_.empty()) {
    throw std::invalid_argument("specs must be a non-empty mapping");
  }
  if (!finite(min_margin_)) {
    throw std::invalid_argument("min_margin must be finite");
  }
  if (confirmations_ < 1) {
    throw std::invalid_argument("confirmations must be >= 1");
  }
  if (max_hamming_ < 0) {
    throw std::invalid_argument("max_hamming must be >= 0");
  }
  if (!finite(loss_timeout_) || loss_timeout_ <= 0.0) {
    throw std::invalid_argument("loss_timeout must be > 0");
  }
  if (!finite(confirmation_window_) || confirmation_window_ <= 0.0) {
    throw std::invalid_argument("confirmation_window must be > 0");
  }
  if (!finite(max_translation_jump_) || max_translation_jump_ <= 0.0) {
    throw std::invalid_argument("max_translation_jump must be > 0");
  }
  if (!finite(max_yaw_jump_) || max_yaw_jump_ <= 0.0) {
    throw std::invalid_argument("max_yaw_jump must be > 0");
  }
  if (!finite(publish_period_) || publish_period_ < 0.0) {
    throw std::invalid_argument("publish_period must be >= 0");
  }
}

GateResult TagGate::evaluate(const std::vector<Detection> & detections, double now)
{
  if (detections.empty()) {
    return reject("NO_TAG");
  }
  if (detections.size() != 1U) {
    return reject("MULTI_TAG");
  }

  const Detection & detection = detections.front();
  const auto dock_it = specs_.find(detection.tag_id);
  if (dock_it == specs_.end()) {
    return reject("UNKNOWN_TAG");
  }
  if (detection.hamming > max_hamming_) {
    return reject("HAMMING");
  }
  if (detection.decision_margin < min_margin_) {
    return reject("LOW_MARGIN");
  }

  if (last_seen_.has_value() && now - *last_seen_ > loss_timeout_) {
    last_accepted_.reset();
    reset_confirmation();
  }
  last_seen_ = now;

  if (is_pose_jump(detection)) {
    return reject("POSE_JUMP");
  }

  const DockSpec & dock = dock_it->second;
  if (confirmed_ && active_tag_id_ != detection.tag_id) {
    reset_confirmation();
  }

  if (!confirmed_) {
    const bool expired = !confirmation_started_.has_value() ||
      now - *confirmation_started_ > confirmation_window_;
    if (active_tag_id_ != detection.tag_id || expired) {
      active_tag_id_ = detection.tag_id;
      confirmation_count_ = 1;
      confirmation_started_ = now;
    } else {
      ++confirmation_count_;
    }

    if (confirmation_count_ < confirmations_) {
      return {false, "CONFIRMING", detection, dock};
    }
    confirmed_ = true;
  }

  if (last_publication_.has_value() && now - *last_publication_ < publish_period_) {
    return {false, "RATE_LIMITED", detection, dock};
  }

  last_accepted_ = detection;
  last_publication_ = now;
  return {true, "ACCEPTED", detection, dock};
}

std::optional<std::string> TagGate::loss_reason(double now)
{
  if (!last_seen_.has_value() || now - *last_seen_ <= loss_timeout_) {
    return std::nullopt;
  }
  last_seen_.reset();
  last_accepted_.reset();
  reset_confirmation();
  return "TAG_LOST";
}

GateResult TagGate::reject(const std::string & reason)
{
  reset_confirmation();
  return {false, reason, std::nullopt, std::nullopt};
}

void TagGate::reset_confirmation()
{
  active_tag_id_.reset();
  confirmation_count_ = 0;
  confirmation_started_.reset();
  confirmed_ = false;
}

bool TagGate::is_pose_jump(const Detection & detection) const
{
  if (!last_accepted_.has_value() || last_accepted_->tag_id != detection.tag_id) {
    return false;
  }
  const double translation = std::hypot(
    detection.x - last_accepted_->x,
    detection.y - last_accepted_->y);
  const double yaw_delta = wrapped_yaw_delta(detection.yaw - last_accepted_->yaw);
  return translation > max_translation_jump_ || std::abs(yaw_delta) > max_yaw_jump_;
}

}  // namespace demo2_apriltag_docking_cpp::core
